File: 2021/day05/zig/solution.zig
const std = @import("std");

const Line = struct {
    x1: i32,
    y1: i32,
    x2: i32,
    y2: i32,
};

fn sign(x: i32) i32 {
    if (x > 0) return 1;
    if (x < 0) return -1;
    return 0;
}

fn countOverlaps(lines: []const Line, include_diagonals: bool, allocator: std.mem.Allocator) !u32 {
    var grid = std.AutoHashMap(u64, u16).init(allocator);
    defer grid.deinit();

    for (lines) |line| {
        const dx = sign(line.x2 - line.x1);
        const dy = sign(line.y2 - line.y1);

        // Skip diagonals in part 1
        if (!include_diagonals and dx != 0 and dy != 0) {
            continue;
        }

        var x = line.x1;
        var y = line.y1;
        while (true) {
            const key = (@as(u64, @intCast(@as(u32, @bitCast(x)))) << 32) | @as(u64, @intCast(@as(u32, @bitCast(y))));
            const entry = try grid.getOrPut(key);
            if (!entry.found_existing) {
                entry.value_ptr.* = 0;
            }
            entry.value_ptr.* += 1;

            if (x == line.x2 and y == line.y2) break;
            x += dx;
            y += dy;
        }
    }

    var count: u32 = 0;
    var iter = grid.valueIterator();
    while (iter.next()) |value| {
        if (value.* >= 2) {
            count += 1;
        }
    }
    return count;
}

fn parseInt(str: []const u8) !i32 {
    const trimmed = std.mem.trim(u8, str, " \t\r\n");
    return try std.fmt.parseInt(i32, trimmed, 10);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const file_size = (try file.stat()).size;
    const buffer = try allocator.alloc(u8, file_size);
    defer allocator.free(buffer);

    _ = try file.readAll(buffer);

    const content = std.mem.trim(u8, buffer, &std.ascii.whitespace);

    // Parse lines
    var lines = std.ArrayListUnmanaged(Line){};
    defer lines.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r\n");
        if (trimmed.len == 0) continue;

        // Parse "x1,y1 -> x2,y2"
        var parts = std.mem.splitSequence(u8, trimmed, " -> ");
        const start = parts.next() orelse continue;
        const end = parts.next() orelse continue;

        var start_parts = std.mem.splitScalar(u8, start, ',');
        const x1 = try parseInt(start_parts.next() orelse continue);
        const y1 = try parseInt(start_parts.next() orelse continue);

        var end_parts = std.mem.splitScalar(u8, end, ',');
        const x2 = try parseInt(end_parts.next() orelse continue);
        const y2 = try parseInt(end_parts.next() orelse continue);

        try lines.append(allocator, Line{ .x1 = x1, .y1 = y1, .x2 = x2, .y2 = y2 });
    }

    const part1 = try countOverlaps(lines.items, false, allocator);
    const part2 = try countOverlaps(lines.items, true, allocator);

    std.debug.print("Part 1: {d}\n", .{part1});
    std.debug.print("Part 2: {d}\n", .{part2});
}

File: 2021/day05/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;

sub sign {
    my ($x) = @_;
    return $x > 0 ? 1 : $x < 0 ? -1 : 0;
}

sub parse_input {
    my $dir = dirname(__FILE__);
    my $input_path = "$dir/../input.txt";

    open(my $fh, '<', $input_path) or die "Cannot open $input_path: $!";

    my @lines;
    while (my $line = <$fh>) {
        chomp $line;
        next if $line eq '';

        if ($line =~ /(\d+),(\d+)\s*->\s*(\d+),(\d+)/) {
            push @lines, [$1, $2, $3, $4];
        }
    }

    close($fh);
    return @lines;
}

sub count_overlaps {
    my ($lines_ref, $include_diagonals) = @_;
    my %grid;

    for my $line (@$lines_ref) {
        my ($x1, $y1, $x2, $y2) = @$line;
        my $dx = sign($x2 - $x1);
        my $dy = sign($y2 - $y1);

        # Skip diagonals in part 1
        next if (!$include_diagonals && $dx != 0 && $dy != 0);

        my ($x, $y) = ($x1, $y1);
        while (1) {
            $grid{"$x,$y"}++;
            last if ($x == $x2 && $y == $y2);
            $x += $dx;
            $y += $dy;
        }
    }

    my $count = 0;
    for my $v (values %grid) {
        $count++ if $v >= 2;
    }
    return $count;
}

sub part1 {
    my @lines = @_;
    return count_overlaps(\@lines, 0);
}

sub part2 {
    my @lines = @_;
    return count_overlaps(\@lines, 1);
}

my @lines = parse_input();
print "Part 1: ", part1(@lines), "\n";
print "Part 2: ", part2(@lines), "\n";
